compiler/rustc_target/src/spec/tests/tests_impl.rs

```diff
@@ -109,20 +112,63 @@ impl Target {
             );
         }
 
-        assert!(
-            (self.pre_link_objects_self_contained.is_empty()
-                && self.post_link_objects_self_contained.is_empty())
-                || self.link_self_contained != LinkSelfContainedDefault::False
-        );
+        if self.link_self_contained == LinkSelfContainedDefault::False {
+            assert!(
+                self.pre_link_objects_self_contained.is_empty()
+                    && self.post_link_objects_self_contained.is_empty()
+            );
+        }
 
         // If your target really needs to deviate from the rules below,
         // except it and document the reasons.
         // Keep the default "unknown" vendor instead.
         assert_ne!(self.vendor, "");
+        assert_ne!(self.os, "");
         if !self.can_use_os_unknown() {
             // Keep the default "none" for bare metal targets instead.
             assert_ne!(self.os, "unknown");
         }
+
+        // Check dynamic linking stuff
+        // FIXME: How does bpf do dynamic linking?
+        if self.os == "none" && self.arch != "bpf" {
+            assert!(!self.dynamic_linking);
+        }
+        if self.only_cdylib
+            || self.crt_static_allows_dylibs
+            || !self.late_link_args_dynamic.is_empty()
+        {
+            assert!(self.dynamic_linking);
+        }
+        // Apparently PIC was slow on wasm at some point, see comments in wasm_base.rs
+        if self.dynamic_linking && !(self.is_like_wasm && self.os != "emscripten") {
+            assert_eq!(self.relocation_model, RelocModel::Pic);
+        }
+        // PIEs are supported but not enabled by default with linuxkernel target.
+        if self.position_independent_executables && !triple.ends_with("-linuxkernel") {
+            assert_eq!(self.relocation_model, RelocModel::Pic);
+        }
+        // FIXME: Why do these target enable PIC if they support neither
+        // dynamic libraries nor position-independent executables?
+        if self.relocation_model == RelocModel::Pic
+            && !(self.os == "uefi"
+                || self.os == "l4re"
+                || self.arch == "avr"
+                || self.is_like_osx && self.os != "macos")
```

Comment:
cc @dvdhrm @Ayush1325 @atopia @dylanmckay @Thog 
Why do these targets enable position-independent code (PIC) even if then support neither dynamic linking nor position-independent executables?
They could use more optimizable `RelocModel::Static` instead.

Reply:
@petrochenkov  I could be wrong, but UEFI works with PE DLLs (essentially; you have to make sure your code is position independent). You can find references to it in the following places:
1. https://www.rodsbooks.com/efi-programming/hello.html
2. https://forum.osdev.org/viewtopic.php?f=1&t=43326&start=0

Reply:
@petrochenkov Ok, so with some discussion in EDK2 community, it seems that PIC shouldn't be needed. EDK2 compiles code with `-nopic` flag by default.

Reply:
@petrochenkov Good point, I'm pretty sure it is an oversight and was just never changed from the default. As I'm about to leave for vacation I've raised the point with @ManuelvOK who has taken over maintenance of the Rust port for L4Re.

Reply:
> cc @dvdhrm @Ayush1325 @atopia @dylanmckay @Thog Why do these targets enable position-independent code (PIC) even if they support neither dynamic linking nor position-independent executables? They could use more optimizable `RelocModel::Static` instead.

The UEFI target compiles into PE binaries, so there is no reason to enable PIC. All that is needed is the relocation table, and this is generated by default for PE binaries, so it is safe to drop this.

I think the PIC annotation is a remnant from the way `gnu-efi` compiles UEFI binaries on most linux machines. They use ELF applications with a PE-loader, and thus they *require* PIC as they cannot correctly fill in the PE relocation data. However, rust is not affected by this `gnu-efi`'ism in any way, so this is likely just an oversight.

However, I am unsure what the `RelocModel` exactly refers to. The LLVM values are not annotated with an explanation, and there are some hints that `Static` means `non-relocatable`. If that is the case, this would be wrong. The code must be relocatable for UEFI, since just like Windows it relies on moving the code if the preferred base address is taken. But there is no requirement for **position-independence**. Ultimately, all UEFI needs is to make sure the relocation-table is not stripped from the final PE binary.

Reply:
L4Re does not support position-independent executables. It should be save to remove the according line.

Reply:
I think Apple enforce PIC with macOS, iOS, tvOS and watchOS. not sure for embededOS (not even sure if Rust have any target for that yet?)

Reply:
Okay, the question about Apple targets resolved itself because dynamic linking support was added to them in https://github.com/rust-lang/rust/pull/100636.

As for other targets, I've changed their relocation model to static.
If something breaks, on UEFI in particular, I guess you'll notice it soon :)